core/internal/runhistoryreader/parquet/iterator/row_iterator.go

```diff
@@ -0,0 +1,391 @@
+// it represents a single metric for a history row.
+// Where Key is the metric name and Value is the metric value.
+type KeyValuePair struct {
+	Key   string
+	Value any
+}
+
+// KeyValueList is a list of KeyValuePairs which represent a single history row.
+type KeyValueList []KeyValuePair
+
+// RowIterator is the primary interface to iterate over history rows
+// for a runs history parquet file.
+type RowIterator interface {
+	// Next advances the iterator to the next value.
+	Next() (bool, error)
+	// Value returns the current value of the iterator.
+	Value() KeyValueList
+	// Release releases the resources used by the iterator.
+	Release()
+}
+
+func selectColumns(schema *schema.Schema, columns []string) ([]int, error) {
+	if len(columns) == 0 {
+		indices := make([]int, schema.NumColumns())
+		for i := 0; i < len(indices); i++ {
+			indices[i] = i
+		}
+		return indices, nil
+	}
+
+	foundMap := map[string]struct{}{}
+
+	var indices []int
+	for i := 0; i < schema.NumColumns(); i++ {
+		col := schema.Column(i)
+		for _, selected := range columns {
+			if col.ColumnPath()[0] == selected {
+				indices = append(indices, i)
+				foundMap[selected] = struct{}{}
+			}
+		}
+	}
+
+	// Make sure that selected columns were found.
+	for _, selected := range columns {
+		if _, ok := foundMap[selected]; !ok {
+			return nil, fmt.Errorf("column %s selected but not found", selected)
+		}
+	}
+
+	return indices, nil
+}
+
+func selectRowGroups(pf *file.Reader, config IteratorConfig) ([]int, error) {
+	indices := []int{}
+
+	// filter row groups that are outside of the requested step range
+	metadata := pf.MetaData()
+	for i := 0; i < len(metadata.RowGroups); i++ {
+		minValue, maxValue, err := getColumnRangeFromStats(pf.MetaData(), i, config.Key)
+		if err != nil {
+			return nil, err
+		}
+
+		if (config.Max == 0 || minValue < config.Max) && (config.Min == math.MaxInt64 || maxValue >= config.Min) {
+			indices = append(indices, i)
+		}
+	}
+
+	return indices, nil
+}
+
+// Iterator uses recordIterators
+// to iterate over arrow.RecordBatches in the given parquet file
+type Iterator struct {
+	ctx    context.Context
+	reader *pqarrow.FileReader
+	keys   []string
+	config IteratorConfig
+
+	needPage        bool
+	rowGroupIndices []int
+
+	recordReader pqarrow.RecordReader
+	current      RowIterator
+}
+
+type IteratorConfig struct {
+	Key string
+	Min float64
+	Max float64
+}
+
+func (c *IteratorConfig) Apply(opts ...RowIteratorOption) {
+	for _, opt := range opts {
+		if opt == nil {
+			continue
+		}
+		opt(c)
+	}
+}
+
+type RowIteratorOption func(*IteratorConfig)
+
+func WithHistoryPageRange(page HistoryPageParams) RowIteratorOption {
+	return func(c *IteratorConfig) {
+		c.Key = StepKey
+		c.Min = float64(page.MinStep)
+		c.Max = float64(page.MaxStep)
+	}
+}
+
+func WithEventsPageRange(page EventsPageParams) RowIteratorOption {
+	return func(c *IteratorConfig) {
+		c.Key = TimestampKey
+		c.Min = page.MinTimestamp
+		c.Max = page.MaxTimestamp
+	}
+}
```

Comment:
Why not

```go
// Let the fields be private, so that the constructors below must be used.
type IteratorConfig struct {
	key string
	min float64
	max float64
}

// No need for Params classes since the parameters are so simple.
// Just follow the convention of passing min before max.
//
// RowIteratorOption doesn't seem appropriate as the options are mutually
// exclusive and not meant to be combined.
func HistoryPageRange(minStep, maxStep int64) IteratorConfig
func EventsPageRange(minTimestamp, maxTimestamp float64) IteratorConfig
```